modules/codegen/src/main/scala/com/twilio/guardrail/generators/Http4sServerGenerator.scala

```diff
@@ -554,6 +574,21 @@ object Http4sServerGenerator {
         )
       })
 
+    def createHttp4sHeaders(headers: List[Header[ScalaLanguage]]): (List[Pat.Var], Term.Name, List[Defn.Val]) = {
+      val (names, definitions) = headers.map {
+        case Header(name, required, _, termName) =>
+          val nameLiteral = Lit.String(name)
+          val headerName  = Term.Name(s"${name.toCamelCase}Header")
+          val pattern     = Pat.Var(headerName)
+          val v           = if (required) q"List(Header($nameLiteral, $termName))" else q"$termName.map(Header($nameLiteral,_)).toList"
+          (headerName, q"val $pattern = $v")
+      }.unzip
+      val patterns    = headers.map(header => Pat.Var(header.term))
+      val headersTerm = Term.Name("responseHeaders")
```

Comment:
This is unfortunately in the same namespace as user-controlled values. Could you alter the http4s server generator to be like
```diff
-            case LoginUserResponse.Ok(value, xRateLimit, xExpiresAfter) =>
-              val xRateLimitHeader = xRateLimit.map(Header("X-Rate-Limit", _)).toList
-              val xExpiresAfterHeader = xExpiresAfter.map(Header("X-Expires-After", _)).toList
-              val responseHeaders = List(xRateLimitHeader, xExpiresAfterHeader).flatten
-              loginUserOkEntityResponseGenerator(value, responseHeaders: _*)(F, loginUserOkEncoder)
+            case resp: LoginUserResponse.Ok =>
+              val xRateLimitHeader = resp.xRateLimit.map(Header("X-Rate-Limit", _)).toList
+              val xExpiresAfterHeader = resp.xExpiresAfter.map(Header("X-Expires-After", _)).toList
+              val responseHeaders = List(xRateLimitHeader, xExpiresAfterHeader).flatten
+              loginUserOkEntityResponseGenerator(resp.value, responseHeaders: _*)(F, loginUserOkEncoder)
```
? That would completely disambiguate values we control from values the user controls. This is how the other generators work (I checked by diffing the output from current master against the output from this branch)

Reply:
Thanks, great suggestion! It also makes the code much simpler.

Reply:
Generated client code is solved differently thought: we append `Header` suffix to each header such that we can be sure that are no conflicts.

Reply:
The edge case would be if I had two parameters, `foo` and `fooHeader`, then suffixed both with `Header` -- I'd then get a conflict.

Reply:
We've already had cases like this, where someone has a data model named `Request` which conflicts with http4s' built-in types. I'm staying extra vigilant until proper tests for unconstrained terms/types have been written.

Reply:
Ahh, I see. Do you have any suggestion?

Reply:
The resulting diff of https://github.com/twilio/guardrail/pull/382#discussion_r322992570 would solve the problem, as all access would be constrained by the `resp` object.

Reply:
That's already done for the server side. What I'm talking about is a client side where we currently generate something like this: 
```scala
case Ok(resp) =>
        val xOptionalHeaderHeader =
          resp.headers.get("x-optional-header".ci).map(_.value)
        val result =
          for (xRequiredHeaderHeader <- EitherT.fromEither[F](
                 resp.headers
                   .get("x-required-header".ci)
                   .map(_.value)
                   .toRight(
                     ParseFailure(
                       "Missing required header.",
                       "HTTP header 'X-Required-Header' is not present."
                     )
                   )
               ))
            yield
              GetFooResponse.Ok(xRequiredHeaderHeader, xOptionalHeaderHeader)
        result.value.flatMap({
          case Right(v) =>
            F.pure(v)
          case Left(e) =>
            F.raiseError(e)
        })
```

Reply:
Oh, sorry, I misunderstood. I took that diff from the server generator. I think you're right for the client generator, other than currying and partially applying the constructor based on whether the field is required or not. Unsure what it would look like, but it would be great to use applicative here to avoid bindings.

Reply:
Implemented even for http4s client.
```
case Ok(resp) =>
        F.map2(parseRequiredHeader(resp, "x-required-header"), parseOptionalHeader(resp, "x-optional-header"))(GetFooResponse.Ok)
```

It uses two helpers methods `parseRequiredHeader` and `parseOptionalHeader` that are defined directly in the generated `Client` class. Perhaps there is a better place for it?

Reply:
@hanny24 A challenge here is that for more than 22 parameters, even this approach will start failing. I mistakenly assumed it would be as easy as `Ok.apply _ curried`, but it appears that that machinery is limited to 22 parameters _as well_ for some reason.

The only thing I can think of that gets us close to safety here is of the structure:

```scala
case class Foo(
  a1: Int, a2: Int, a3: Int, a4: Int, a5: Int, a6: Int, a7: Int, a8: Int, a9: Int, a10: Int, a11: Int, a12: Int, a13: Int, a14: Int, a15: Int, a16: Int, a17: Int, a18: Int, a19: Int, a20: Int, a21: Int, a22: Int,
  a23: Int, a24: Int, a25: Int, a26: Int, a27: Int, a28: Int, a29: Int
)

Applicative[Option].ap7(
  Applicative[Option].map22(
    Option(1), Option(2), Option(3), Option(4), Option(5), Option(6), Option(7),
    Option(8), Option(9), Option(10), Option(11), Option(12), Option(13),
    Option(14), Option(15), Option(16), Option(17), Option(18), Option(19),
    Option(20), Option(21), Option(22)
  )(
    (
      a1: Int, a2: Int, a3: Int, a4: Int, a5: Int, a6: Int, a7: Int, a8: Int,
      a9: Int, a10: Int, a11: Int, a12: Int, a13: Int, a14: Int, a15: Int,
      a16: Int, a17: Int, a18: Int, a19: Int, a20: Int, a21: Int, a22: Int
    ) =>
      Foo(
        a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16,
        a17, a18, a19, a20, a21, a22,
        _: Int, _: Int, _: Int, _: Int, _: Int, _: Int, _: Int
      )
  )
)(
  Option(23), Option(24), Option(25), Option(26), Option(27), Option(28),
  Option(29)
)
```
but that requires _way_ more juggling than I'd like to take on currently. You're welcome to, if you'd like, otherwise just assert that if there are more than 22 parameters for the resulting `Response` container, fail the codegen and link to this thread (both in the error message and a comment nearby).

Reply:
Ok, the build now succeeds. Once the PR is approved I can squash all changes.

Reply:
I just pushed a couple more commits, pushing the validation logic into `build` so the validation happens close to the consumption, to prevent it from getting lost.

I'm onboard merging with this, if you have no objections.

Thank you for your patience through this whole process

Reply:
I'm OK with your commits. Feel free to merge it. Thanks!